log4j-slf4j20-impl/src/main/java/org/apache/logging/slf4j/Log4jMDCAdapter.java

```diff
@@ -27,57 +29,229 @@
+        }
+
+        @Override
+        public void clear() {
+            ThreadContext.clearMap();
+        }
+
+        @Override
+        public Map<String, String> getCopyOfContextMap() {
+            return ThreadContext.getContext();
+        }
+
+        @Override
+        public void setContextMap(final Map<String, String> map) {
+            ThreadContext.clearMap();
+            ThreadContext.putAll(map);
+        }
+
+        @Override
+        public void pushByKey(String key, String value) {
+            // NOP
+        }
+
+        @Override
+        public String popByKey(String key) {
+            // NOP
+            return null;
+        }
+
+        @Override
+        public Deque<String> getCopyOfDequeByKey(String key) {
+            // NOP
+            return null;
+        }
+
+        @Override
+        public void clearDequeByKey(String key) {
+            // NOP
+        }
+    }
+
+    /**
+     * Wires the top of the stack to the MDC current value
+     */
+    private static final class ExtendedMDCAdaptor extends StandardMDCAdaptor {
+
+        private final ThreadLocalMapOfStacks   threadLocalMapOfDeques = new ThreadLocalMapOfStacks();
+
+        @Override
+        public void clear() {
+            super.clear();
+            threadLocalMapOfDeques.clear();
+        }
+
+        @Override
+        public void pushByKey(String key, String value) {
+            threadLocalMapOfDeques.pushByKey(key, value);
+            put(key, value);
+        }
+
+        @Override
+        public String popByKey(String key) {
+            String value = threadLocalMapOfDeques.popByKey(key);
+            String head = threadLocalMapOfDeques.peekByKey(key);
+            if (head != null) {
+                put(key, head);
+            }
+            return value;
+        }
+
+        @Override
+        public Deque<String> getCopyOfDequeByKey(String key) {
+            return threadLocalMapOfDeques.getCopyOfDequeByKey(key);
+        }
+
+        @Override
+        public void clearDequeByKey(String key) {
+            threadLocalMapOfDeques.clearDequeByKey(key);
+        }
+    }
+
+    /**
+     * A simple implementation of ThreadLocal backed Map containing values of type
+     * Deque<String>. This class is inspired from SLF4J version 2.0.0
+     */
+    private static final class ThreadLocalMapOfStacks {
+
+        private final ThreadLocal<Map<String, Deque<String>>> tlMapOfStacks = new ThreadLocal<>();
+
+        void pushByKey(String key, String value) {
+            if (key == null)
+                return;
+
+            Map<String, Deque<String>> map = tlMapOfStacks.get();
+
+            if (map == null) {
+                map = new HashMap<>();
+                tlMapOfStacks.set(map);
+            }
+
+            Deque<String> deque = map.get(key);
+            if (deque == null) {
+                deque = new ArrayDeque<>();
+            }
+            deque.push(value);
+            map.put(key, deque);
```

Comment:
We could issue a warning if the `key` is already used in the classical MDC (`ThreadContext.get() != null`) (cf. [SLF4J-531](https://jira.qos.ch/browse/SLF4J-531) and return.

There is really no reason to allow users to use the same `key` in both the classical and stack-based MDC.